packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -270,69 +275,24 @@ export default class RequestClient {
   /**
    * @param {UppyFile} file
    */
-  async connectToServerSocket (file, queue) {
-    return new Promise((resolve, reject) => {
-      const token = file.serverToken
-      const host = getSocketHost(file.remote.companionUrl)
-      const socket = new Socket({
-        target: `${host}/api/${token}`,
-        autoOpen: false,
-      })
-      const eventManager = new EventManager(this.uppy)
-
-      let queuedRequest
-
-      eventManager.onFileRemove(file.id, () => {
+  async connectToServerSocket(file) {
+    const token = file.serverToken
+    const host = getSocketHost(file.remote.companionUrl)
+    const socket = new Socket({ target: `${host}/api/${token}`, autoOpen: false })
+    const { capabilities } = this.uppy.getState()
+    const deferred = defer()
```

Comment:
why do we need defer here? can't we just use the [`new Promise` constructor](https://github.com/transloadit/uppy/pull/4734/files#diff-99b347c8acf27f8707896e4d330a41671d6bb9257398e4c265f27f77ad434a83R339)? I thought defer was considered an anti-pattern

Reply:
Mostly because we need to return the promise itself in the queue add:

```js
    this.uppy.queue.add(() => {
      if (file.isPaused) {
        socket.send('pause', {})
      } else {
        socket.open()
      }
      return deferred.promise
    }, { signal: controller.signal })
```